src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -1,21 +1,355 @@
+// with at least one active deals and deals cannot be updated
+// an alternative proposal is to account for power based on active deals
+// an improvement proposal is to allow storage deal update in a sector
+
+// TODO: decide whether declared faults sectors should be
+// penalized in the same way as undeclared sectors and how
+
+// SubmitPoSt Workflow:
+// - Verify PoSt Submission
+// - Process ProvingSet.GetOnes()
+//   - State Transitions
+//     - Committed -> Active and credit power
+//     - Recovering -> Active and credit power
+//   - Process Active Sectors (pay miners)
+// - Process ProvingSet.GetZeros()
+//	   - increment FaultCount
+//     - clear Sector and slash pledge collateral if count > MAX_CONSECUTIVE_FAULT_COUNT
+// - Process Expired Sectors (settle deals and return storage collateral to miners)
+//     - State Transition
+//       - Failing / Recovering / Active / Committed -> Cleared
+//     - Remove SectorNumber from Sectors, SectorStates, ProvingSet
+func (sm *StorageMinerActor_I) SubmitPoSt(postSubmission poster.PoStSubmission) {
+	// Verify correct PoSt Submission
+	isPoStVerified := sm.verifyPoStSubmission(postSubmission)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitPoSt again
+		// TODO: proper failure
+		panic("TODO")
+	}
+
+	// The proof is verified, process ProvingSet.GetOnes():
+	// ProvingSet.GetOnes() contains SectorCommitted, SectorActive, SectorRecovering
+	// ProvingSet itself does not store states, states are all stored in SectorStates
+	var powerUpdate uint64 = 0
+
+	for _, sectorNo := range sm.ProvingSet_.GetOnes() {
+		state := sm.SectorStates()[sectorNo]
+		switch state.StateNumber {
+		case SectorCommittedStateNo:
+			sm.activateSector(sectorNo)
+			powerUpdate = powerUpdate + sm.Info().SectorSize()
+		case SectorRecoveringStateNo:
+			// Note: SectorState.FaultCount is also reset to zero here
+			sm.activateSector(sectorNo)
+			powerUpdate = powerUpdate + sm.Info().SectorSize()
+		case SectorActiveStateNo:
+			// Process payment in all active deals
+			// Note: this must happen before marking sectors as expired.
+			// TODO: Pay miner in a single batch message
+			// SendMessage(sma.ProcessStorageDealsPayment(sm.Sectors()[sectorNumber].DealIDs()))
+		default:
+			// TODO: proper failure
+			panic("Invalid sector state in ProvingSet.GetOnes()")
+		}
+	}
+
+	// Process ProvingSet.GetZeros()
+	// ProvingSet.GetZeros() contains SectorFailing
+	// SectorRecovering is Proving and hence will not be in GetZeros()
+	// heavy penalty if Failing for more than or equal to MAX_CONSECUTIVE_FAULT_COUNT
+	// otherwise increment FaultCount in SectorStates()
+	for _, sectorNo := range sm.ProvingSet_.GetZeros() {
+		state := sm.SectorStates()[sectorNo]
+		switch state.StateNumber {
+		case SectorFailingStateNo:
+			sm.checkMFC(sectorNo)
+		default:
+			// TODO: proper failure
+			panic("Invalid sector state in ProvingSet.GetZeros")
+		}
+	}
+
```

Comment:
I would add a note here (as a comment) that expiration needs to happen in the same flow, _after_ states change to Failing or Active (ie Committed and Recovering pass through those before expiring)

Reply:
addressed in 435ab9f